snuba/subscriptions/worker.py

```diff
@@ -35,21 +42,40 @@ def __init__(
 
     def process_message(
         self, message: Message[Tick]
-    ) -> Optional[Sequence[SubscriptionResult]]:
-        results: MutableSequence[SubscriptionResult] = []
-
+    ) -> Optional[Sequence[SubscriptionResultFuture]]:
+        # Schedule all of the queries (tasks) associated with this tick message
+        # for (asynchronous) execution.
+        # NOTE: There is no backpressure here (tasks will be submitted to the
+        # executor as quickly as they are received from the broker) but there
+        # *is* a concurrency limit on the executor. If a much larger number of
+        # queries are scheduled to execute than there is capacity to execute,
+        # waiting for them to complete during ``flush_batch`` may exceed the
+        # consumer poll timeout (or session timeout during consumer
+        # rebalancing) and cause the entire batch to be have to be replayed.
         tick = message.payload
-        for task in self.__schedulers[message.partition.index].find(tick.timestamps):
-            results.append(
-                SubscriptionResult(task, self.__executor.execute(task, tick).result())
-            )
+        return [
+            SubscriptionResultFuture(task, self.__executor.execute(task, tick))
+            for task in self.__schedulers[message.partition.index].find(tick.timestamps)
+        ]
 
-        return results
+    def flush_batch(self, batch: Sequence[Sequence[SubscriptionResultFuture]]) -> None:
+        # Map over the batch, converting the ``SubscriptionResultFuture`` to a
+        # ``SubscriptionResult``. This will block and wait for any unfinished
+        # queries to complete, and raise any exceptions encountered during
+        # query execution. Either the entire batch succeeds, or the entire
+        # batch fails.
+        results = [
+            SubscriptionResult(task, future.result())
+            for task, future in itertools.chain.from_iterable(batch)
+        ]
 
-    def flush_batch(self, batch: Sequence[Sequence[SubscriptionResult]]) -> None:
-        for results in batch:
-            for result in results:
-                self.__producer.produce(self.__topic, result).result()
+        # Produce all of the subscription results asynchronously and wait for
+        # them to complete. Again, either the entire batch succeeds, or the
+        # entire batch fails.
+        for future in as_completed(
+            [self.__producer.produce(self.__topic, result) for result in results]
+        ):
+            future.result()
```

Comment:
This is still pretty inefficient. It wastes time blocking waiting for results (from the executor, the producer, not to mention the broker on commit) that could be better spent processing messages from the next batch, and making this be an all-or-nothing operation instead of allowing for a partial commit means that we have to buffer producing the results to the topic longer than necessary, adding latency to the system. I think this is about as good as it will get without a substantial redesign of the consumer though (which will probably have to happen at some point, but I am trying to avoid.)